### ISSUE TITLE
nspawn breaks pm_request

### ISSUE THREAD

**Issue by nim-nim** [rpm-software-management/mock#218]
pm_request does not work when nspawn is enabled, you can see the socket created from the host system, but it is not accessible from within the build environment

mock-1.4.13-1.fc29.noarch (same failure on an el7 host with the el7 mock version)

build target: el7

**Comment by nim-nim** [comment]
This is related to
https://github.com/rpm-software-management/mock/issues/160
and https://github.com/rpm-software-management/rpm/issues/104

**Comment by praiskup** [comment]
I could prepare PR for this; seems to be pretty trivial  - systemd-nspawn overmounts `/run` -> while we are creating the file under `mock_chroot/run`.  What is the proper place to put the socket file then?  I'm thinking about `/var/tmp/` as a compromise;  but do we have to care about concurrency from e.g. multiple `mock --shell` processes?

Also, while am I on it, I'd propose to change the socket mode to 0660, instead of 0775.

**Comment by nim-nim** [comment]
Well, I have no opinion on all of those, from my side, go for it if it fixes the nspawn incompatibility. I do run many mock builds in parallel (via separate mock cfg files which have the same content except for the conf name and filesystem root) but `mock --shell`? almost never.

**Comment by xsuchy** [comment]
I think you have to pass `--bind=` to nspawn with the path to the socket. Or in mock you can utilize bind_mount plugin.

**Comment by praiskup** [comment]
@msuchy
> I think you have to pass --bind= to nspawn 

Do we have to keep it under `/run`?  (just asking your preferences, I'd probably agree that - though complicated - it would be cleaner).  Otherwise, because we bind-mount the parent (root) directory - we shouldn't need that for e.g. `/var/tmp/mock` or say `/var/tmp/mock/pm_request.<PID>.sock`.

**Comment by xsuchy** [comment]
`/run` and other tmpfs mounted directories are problematic. I am not sure if this has been resolved in mock, but the problem was in past that mock bind mount some directory, then we start nspawn, which mounted the tmpfs over th bind mount. I *think* we resolved it by using `--bind`, but I am not sure.
You can have the socket in /run on host, but in chroot I would bind mount it to /var/tmp/

**Comment by keszybz** [comment]
I think on the host you want to use something like `/tmp/mock.pm_request.XXXXXX` with the XXXXXX part filled using random numbers. Predictable names like `/var/tmp/mock` are always problematic because of DOS/conflicts/races.

Please use `--bind`. It's designed for this purpose, and it this is much nicer than relying on some temporary directory not being covered.

On the buildroot side, I think using a path in `/run` or `/tmp` should work equally well, as long as the path is specified using an environment variable. (It would be possible to use a fixed path like `/run/mock.pm_requests`, but that becomes inflexible, for example it is harder to test pm_mock by running it i a host system. With an environment variable specifying the path, it's easier to deploy outside of a fully virtualized env.)

**Comment by nim-nim** [comment]
So FESCO reviewed the possibility on activating pm request on Fedora and the only real sticking point was security:
https://pagure.io/fesco/issue/2004#comment-539329

So, basically, filter the argument list dnf passes to pm request and make sure the result does not escape the mock jail. Unless I’m mistaken (?) making sure dnf does not escape the mock jail means
 * making nspawn work 
 * expose a special dnf config dir in nswpawn which is ro from within the container (so probably not the usual config dir to avoid collinding with the repo files installed by the packages themselves)
 * restricting the dnf pm request calls to this special config dir, ignoring the usual one

**Comment by keszybz** [comment]
Hmm, there's another tricky point: BR can be specified as paths, for file deps (e.g. `/usr/bin/executable`). The plugin must distinguish this from file names (`/some/where/my.rpm`, and in particular `/path/to/mock/root/evil.rpm`). Some heuristics might be necessary here. Alternatively, file deps may simply be not supported by the plugin.

> And probably fork pm request in a safe pm request version since some people want the current unsafe mode being kept available.

Why not just make this a configuration option? I think the default could be the "safe" mode, since allowing dnf to break out of the build root is a very niche need.

**Comment by nim-nim** [comment]
I don't need file deps for my use case so not supporting them and defferring them to
rpm-software-management/rpm#104 would be ok for me

I'm kind of surprised though, can't the dnf mock calls just be executed in a form of container where it does not see the host  filesystem?

**Comment by nim-nim** [comment]
Well I don't need file deps for my use case, so they can be deferred to
https://github.com/rpm-software-management/rpm/issues/104
though I'm sure someone will want to use them eventually.

https://bugzilla.redhat.com/show_bug.cgi?id=1642840

If the dnf pm request calls is restricted by nspawn or something similar to only see and write in the buildroot, the problem goes away.

If that can't be done putting security dnf heuristics mock side is probably a terrible idea, you really need something like:
https://bugzilla.redhat.com/show_bug.cgi?id=1642840

A dnf mode where dnf only performs installs/updates, from the repositories specified in a safe locked-down conf dir, and nothing else.

**Comment by nim-nim** [comment]
Actually, you can do it without heuristics mock side, just test if the file exists on the filesystem, and if `file` says it's a rpm, just remove it from args. That should be fairly simple and foolproof.

**Comment by xsuchy** [comment]
@nim-nim is this still needed? Or Dynamic Build Requires made this obsolete?

**Comment by xsuchy** [comment]
I am going to close this. I do not want to punch more holes into nspawn. and DynamicBuildRequires should solve this use case.
